pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,567 @@
+func (proxier *Proxier) syncProxyRules() error {
+	glog.V(2).Infof("Syncing iptables rules.")
+
+	// lock proxy state so that we can sync rules from the current state
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+
+	// Get iptables-save output so we can check for existing chains and rules.
+	// This will be a map of chain name to chain with rules as stored in iptables-save/iptables-restore
+	existingChains := make(map[utiliptables.Chain]string)
+	// run iptables-save
+	iptablesSaveRaw, err := proxier.iptables.Save(utiliptables.TableNAT)
+	if err != nil { // if we failed to get any rules
+		glog.Errorf("Failed to execute iptable-save, syncing all rules. (%s)", err.Error())
+	} else { // otherwise parse the output
+		existingChains = utiliptables.ParseTableSaveChains(utiliptables.TableNAT, iptablesSaveRaw)
+	}
+
+	// for first line and chains
+	var rules bytes.Buffer
+	// for the actual rules (which should be after the list of chains)
+	var rulesEnd bytes.Buffer
+
+	// write table header
+	rules.WriteString("*nat\n")
+
+	if chain, ok := existingChains[iptablesServicesChain]; ok {
+		rules.WriteString(chain)
+		rules.WriteRune('\n')
+	} else {
+		rules.WriteString(fmt.Sprintf(fmtChain, iptablesServicesChain))
+	}
+
+	// Flush the chains
+	rulesEnd.WriteString(fmt.Sprintf(fmtFlush, iptablesServicesChain))
+
+	newHostChains := []utiliptables.Chain{}
+	newServiceChains := []utiliptables.Chain{}
+
+	//Build rules for services
+	for name, info := range proxier.serviceMap {
+		protocol := strings.ToLower((string)(info.protocol))
+
+		// get chain name
+		svcChain := servicePortToServiceChain(name)
+
+		// Create chain
+		if chain, ok := existingChains[svcChain]; ok {
+			rules.WriteString(chain)
+			rules.WriteRune('\n')
+		} else {
+			rules.WriteString(fmt.Sprintf(fmtChain, svcChain))
+		}
+		// Flush chain
+		rulesEnd.WriteString(fmt.Sprintf(fmtFlush, svcChain))
+
+		hosts := make([]string, 0)
+		hostChains := make([]utiliptables.Chain, 0)
+		for _, ep := range info.endpoints {
+			hosts = append(hosts, ep)
+			hostChains = append(hostChains, servicePortAndEndpointToServiceChain(name, ep))
+		}
+
+		// Ensure we know what chains to flush/remove next time we generate the rules
+		newHostChains = append(newHostChains, hostChains...)
+		newServiceChains = append(newServiceChains, svcChain)
+
+		n := len(hostChains)
+		for i, hostChain := range hostChains {
+			// Create chain
+			if chain, ok := existingChains[utiliptables.Chain(hostChain)]; ok {
+				rules.WriteString(chain)
+				rules.WriteRune('\n')
+			} else {
+				rules.WriteString(fmt.Sprintf(fmtChain, hostChain))
+			}
+			// Flush chain
+			rulesEnd.WriteString(fmt.Sprintf(fmtFlush, hostChain))
+			// Sticky session
+			if info.sessionAffinityType == api.ServiceAffinityClientIP {
+				rulesEnd.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m recent --name %s --rcheck --seconds %d --reap -j %s\n", svcChain, name.String(), hostChain, info.stickyMaxAgeMinutes*60, hostChain))
```

Comment:
Won't this result in interleaved lines?

rcheck
statistic
rcheck
statistic

??

That's not what we want.  We need to check all the rcheck before doing any statistical jumps